File: Impostor/ImpostorGameModel.m
//
//  ImpostorGameModel.m
//  Impostor
//
//  Created by William Entriken on 12/14/13.
//  Copyright (c) 2013 William Entriken. All rights reserved.
//

#import "ImpostorGameModel.h"

@interface ImpostorGameModel()
@property (nonatomic) NSInteger numberOfPlayers;
@property (nonatomic) NSArray *playerRoles; // of NSNumber of PlayerRoles
@property (nonatomic) NSArray *playerWords; // of NSStrings
@property (nonatomic) NSMutableArray *playerEliminated; // of NSNumber of BOOLs
@property (nonatomic) GameStatus gameStatus;
@property (nonatomic) NSInteger playerNumberToStartRound;
@property (nonatomic) NSString *normalWord;
@property (nonatomic) NSString *impostorWord;
@end

@implementation ImpostorGameModel

#pragma mark - ACESSORS FOR INITIALIZERS

- (void)startGameWithNumberOfPlayers:(NSInteger)numPlayers
{
    self.numberOfPlayers = numPlayers;
    int impostorIndex = arc4random() % self.numberOfPlayers;
    
    NSMutableArray *roles = [[NSMutableArray alloc] initWithCapacity:numPlayers];
    NSMutableArray *words = [[NSMutableArray alloc] initWithCapacity:numPlayers];
    self.playerEliminated = [[NSMutableArray alloc] initWithCapacity:numPlayers];
    
    NSURL *jsonURL = [[NSBundle mainBundle] URLForResource:@"gameWords" withExtension:@"json"];
    NSString *jsonString = [NSString stringWithContentsOfURL:jsonURL encoding:NSUTF8StringEncoding error:nil];
    NSArray *allWordSets = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
    int chosenWordSet = arc4random() % allWordSets.count;
    int chosenImpostorWord = arc4random() % 2;
    self.impostorWord = ((NSArray *)allWordSets[chosenWordSet])[chosenImpostorWord];
    self.normalWord = ((NSArray *)allWordSets[chosenWordSet])[1-chosenImpostorWord];
    
    for (int i=0; i<numPlayers; i++) {
        if (i == impostorIndex) {
            [roles addObject:@(PlayerRoleImpostor)];
            [words addObject:self.impostorWord];
        } else {
            [roles addObject:@(PlayerRoleNormalPlayer)];
            [words addObject:self.normalWord];
        }
        [self.playerEliminated addObject:@NO];
    }

    self.playerRoles = [roles copy];
    self.playerWords = [words copy];
    self.gameStatus = GameStatusShowSecretWords;
    self.playerNumberToStartRound = arc4random() % self.numberOfPlayers;
}

- (void)eliminatePlayer:(NSInteger)playerNumber
{
    if ([(NSNumber *)self.playerEliminated[playerNumber] boolValue])
        NSLog(@"ERROR eliminating player that is already eliminated");
    self.playerEliminated[playerNumber] = @YES;
    
    int countOfRemainingImpostors = 0;
    int countOfRemainingNormalPlayers = 0;
    for (int i=0; i<self.numberOfPlayers; i++) {
        if ([(NSNumber *)self.playerEliminated[i] boolValue])
            continue;
        switch ([(NSNumber *)self.playerRoles[i] intValue]) {
            case PlayerRoleImpostor:
                countOfRemainingImpostors++;
                break;
            case PlayerRoleNormalPlayer:
                countOfRemainingNormalPlayers++;
                break;
        }
    }
    if (countOfRemainingImpostors == 0)
        self.gameStatus = GameStatusTheImpostorWasDefeated;
    else if (countOfRemainingNormalPlayers == 1)
        self.gameStatus = GameStatusTheImpostorWon;

    self.playerNumberToStartRound = playerNumber;
    while ([(NSNumber *)self.playerEliminated[self.playerNumberToStartRound] boolValue])
        self.playerNumberToStartRound = (self.playerNumberToStartRound + 1) % self.numberOfPlayers;
}

- (void)doneShowingSecretWords
{
    self.gameStatus = GameStatusTheImpostorRemains;
}

+ (ImpostorGameModel *)game
{
    static ImpostorGameModel *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[ImpostorGameModel alloc] init];
    });
    return sharedInstance;
}

@end

File: Impostor/ImpostorGameModel.swift
//
//  ImpostorGameModel.swift
//  Impostor
//
//  Created by William Entriken on 2/20/16.
//  Copyright © 2016 William Entriken. All rights reserved.
//

import Foundation

enum PlayerRoles {
    case normalPlayer
    case impostor
}

enum GameStatus: Equatable {
    case showSecretWord(Int)
    case theImpostorRemains
    case theImpostorWasDefeated
    case theImpostorWon
}

class ImpostorGameModel: NSObject {
    static var game = ImpostorGameModel()
    
    private(set) var gameStatus = GameStatus.showSecretWord(0)
    private(set) var numberOfPlayers = 0
    private(set) var playerRoles = [PlayerRoles]()
    private(set) var playerEliminated = [Bool]()
    private(set) var playerWords = [String]()
    private(set) var playerNumberToStartRound = 0
    private(set) var normalWord = ""
    private(set) var impostorWord = ""
    
    func startGameWithNumberOfPlayers(_ numPlayers: Int) {
        var allWordSets = [[String]]()
        let jsonURL = Bundle.main.url(forResource: "gameWords", withExtension: "json")!
        let jsonData = try! Data(contentsOf: jsonURL)
        allWordSets = try! JSONSerialization.jsonObject(with: jsonData, options: []) as! [[String]]
        if UserDefaults.standard.integer(forKey: "didIAP") > 0 {
            let moreJsonURL = Bundle.main.url(forResource: "gameWordsMore", withExtension: "json")!
            let moreJsonData = try! Data(contentsOf: moreJsonURL)
            let moreWordSets = try! JSONSerialization.jsonObject(with: moreJsonData, options: []) as! [[String]]
            allWordSets.append(contentsOf: moreWordSets)
        }
        let chosenWordSet = allWordSets.randomElement()!
        normalWord = chosenWordSet[0]
        impostorWord = chosenWordSet[1]
        if Bool.random() == true {
            swap(&normalWord, &impostorWord)
        }
        numberOfPlayers = numPlayers
        playerRoles = [PlayerRoles](repeating: .normalPlayer, count: numberOfPlayers)
        playerWords = [String](repeating: self.normalWord, count: numberOfPlayers)
        playerEliminated = [Bool](repeating: false, count: numberOfPlayers)
        let impostorIndex = Int.random(in: 0..<numberOfPlayers)
        playerRoles[impostorIndex] = .impostor
        playerWords[impostorIndex] = self.impostorWord
        playerNumberToStartRound = Int.random(in: 0..<numberOfPlayers)
        gameStatus = .showSecretWord(0)
    }

    func doneShowingSecretWord(_ playerNumber: Int) {
        if playerNumber == numberOfPlayers - 1 {
            gameStatus = .theImpostorRemains
        } else {
            gameStatus = .showSecretWord(playerNumber + 1)
        }
    }
    
    func eliminatePlayer(_ playerNumber: Int) {
        assert(playerEliminated[playerNumber] == false)
        playerEliminated[playerNumber] = true
        
        let remainingPlayerRoles = zip(playerRoles, playerEliminated).compactMap {$1 ? nil : $0}
        let countOfRemainingImpostors = remainingPlayerRoles.filter({$0 == .impostor}).count
        let countOfRemainingNormalPlayers = remainingPlayerRoles.filter({$0 == .normalPlayer}).count
        if countOfRemainingImpostors == 0 {
            gameStatus = .theImpostorWasDefeated
        } else if countOfRemainingNormalPlayers == 1 {
            gameStatus = .theImpostorWon
        }
        playerNumberToStartRound = playerNumber
        while playerEliminated[playerNumberToStartRound] {
            playerNumberToStartRound = (playerNumberToStartRound + 1) % numberOfPlayers
        }
    }
}
